torch/distributions/constraints.py

```diff
@@ -456,16 +460,44 @@ def check(self, value):
         return _LowerCholesky().check(value) & unit_row_norm
 
 
-class _PositiveDefinite(Constraint):
+class _Square(Constraint):
     """
-    Constrain to positive-definite matrices.
+    Constrain to square matrices.
     """
     event_dim = 2
 
     def check(self, value):
-        # Assumes that the matrix or batch of matrices in value are symmetric
-        # info == 0 means no error, that is, it's SPD
-        return torch.linalg.cholesky_ex(value).info.eq(0).unsqueeze(0)
+        return torch.full(
+            size=value.shape[:-2],
+            fill_value=(value.shape[-2] == value.shape[-1]),
+            dtype=torch.bool,
+            device=value.device
+        )
+
+
+class _Symmetric(_Square):
+    """
+    Constrain to Symmetric square matrices.
+    """
+
+    def check(self, value):
+        # square_check = super().check(value)
+        # if not square_check.all():
+        #    return square_check
+        return value.eq(value.mT).all(-2).all(-1)
+
+
+class _PositiveDefinite(_Symmetric):
+    """
+    Constrain to positive-definite matrices.
+    """
+
+    def check(self, value):
+        sym_check = super().check(value)
+        if not sym_check.all():
+            return sym_check
+        # return torch.linalg.cholesky_ex(value).info.eq(0)
+        return torch.linalg.eigvalsh(value).gt(0).all(-1)
```

Comment:
What's the reason for changing this?

Reply:
This does not seem right to me, as this is noticeably slower than the previous approach.

Reply:
`return torch.linalg.cholesky_ex(value).info.eq(0)` will be the last implementation, but I don't understand why we get CI errors. So I tried several changes in the code. I will revert it.